File: iOS/Sources/Beagle/Sources/Common/Observable/Observable.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

public protocol Observer: AnyObject {
    func didChangeValue(_ value: Any?)
}

public protocol ObservableProtocol: AnyObject {
    associatedtype Value

    var observers: [Observer] { get }

    func addObserver(_ observer: Observer)
    func deleteObserver(_ observer: Observer)

    var value: Value { get set }
}

public class Observable<T>: ObservableProtocol {

    public var observers: [Observer] = []
    
    public var value: T {
        didSet { notifyObservers() }
    }
    
    public init(value: T) {
        self.value = value
    }
    
    public func addObserver(_ observer: Observer) {
        guard !isAlreadyObserving(observer) else { return }

        observers.append(observer)
    }
    
    public func deleteObserver(_ observer: Observer) {
        guard let index = observers.firstIndex(where: { $0 === observer }) else {
            return
        }
        observers.remove(at: index)
    }

    private func notifyObservers() {
        observers.forEach {
            $0.didChangeValue(value)
        }
    }

    private func isAlreadyObserving(_ observer: Observer) -> Bool {
        return observers.contains { $0 === observer }
    }
}

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/form/observer/Observable.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.components.form.observer

import br.com.zup.beagle.android.components.form.core.Constants
import java.util.Vector

@Deprecated(Constants.FORM_DEPRECATED_MESSAGE)
class Observable<T> {
    private var changed = false
    private val obs: Vector<Observer<T>> = Vector()
    private var currentValue: T? = null

    @Synchronized
    fun addObserver(o: Observer<T>) {
        if (!obs.contains(o)) {
            obs.addElement(o)
        }

        if (hasChanged()) {
            currentValue?.let {
                notifyObservers(it)
            }
        }
    }

    @Synchronized
    fun deleteObserver(o: Observer<T>) {
        obs.removeElement(o)
    }

    fun notifyObservers(arg: T) {
        currentValue = arg
        setChanged()
        val arrLocal: Array<Any>

        synchronized(this) {
            if (!changed)
                return
            arrLocal = obs.toArray()
            clearChanged()
        }

        for (i in arrLocal.indices.reversed())
            (arrLocal[i] as Observer<T>).update(this, arg)
    }

    @Synchronized
    fun deleteObservers() {
        obs.removeAllElements()
    }

    @Synchronized
    private fun setChanged() {
        changed = true
    }

    @Synchronized
    private fun clearChanged() {
        changed = false
    }

    @Synchronized
    fun hasChanged(): Boolean {
        return changed
    }

    @Synchronized
    fun countObservers(): Int {
        return obs.size
    }

    @Synchronized
    fun getCurrentValue(): T? = currentValue
}
